### PR TITLE
fix(ecs-patterns): handle desired task count being set to 0

### LINKED ISSUE #4719: QueueProcessingFargateService unable to set desiredCount to zero
I was trying to create a QueueProcessingFargateService using the ecs-patterns construct library. 
I only want to start a Fargate Service if there is something waiting in the SQS queue.

Using the QueueProcessingServiceBaseProps it is not possible to create a Fargate Service that has a desiredTaskCount of 0, therefore, there will always be one service running even if no messages are in the queue.

I checked and the value 0 is overwritten and the desiredCount is always set to 1. 

Line 35 in queue-processing-service-base.js
`this.desiredCount = props.desiredTaskCount || 1;` 

Is this the desired effect?

### Environment

  - **CDK CLI Version:** 1.14.0 (build 261a1bf) 
  - **Module Version:**   "@aws-cdk/aws-ecs-patterns": "^1.14.0",
  - **OS:**  OSX Mojave
  - **Language:** TypeScript

### PR DESCRIPTION
**Pull request by nmussy**
Allow desired task count to be set to `0` for the following (defaults to `1` when the property is `undefined`):
* `QueueProcessingServiceBase.desiredTaskCount` property, affects:
   * `QueueProcessingEc2Service`
   * `QueueProcessingFargateService`
Note: If desired task count is set to 0, maxScalingCapacity must be defined with a value greater than 0.

Throw exception when desired task count is set to `0` for the following (defaults to `1` when the property is `undefined`):
* `ApplicationLoadBalancedServiceBase.desiredCount` property, affects:
   * `ApplicationLoadBalancedEc2Service` 
   * `ApplicationLoadBalancedFargateService`
* `NetworkLoadBalancedServiceBase.desiredCount` property, affects:
   * `NetworkLoadBalancedEc2Service`
   * `NetworkLoadBalancedFargateService`
* `ScheduledTaskBase.desiredTaskCount` property, affects:
   * `ScheduledEc2Task`
   * `ScheduledFargateTask`

Fixes #4719

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by nmussy** [comment]
@pkandasamy91 Your title is incorrect, this PR doesn't only cover `queueprocessingservice`, but also `ApplicationLoadBalanced` and `NetworkLoadBalanced` services

**Comment by nmussy** [comment]
@pkandasamy91 We don't have to allow 0, seeing the anti-patterns you described in the issue.

This problem now is that is user isn't aware that they are not able to set a capacity to 0, and the CDK will silently bump it to 1. The way I see it, we can go about it two ways:

* Allow 0, warn the user
* Document the minimum value, throw if set to 0

What do you think?

**Comment by pkandasamy91** [comment]
> @pkandasamy91 We don't have to allow 0, seeing the anti-patterns you described in the issue.
> 
> This problem now is that is user isn't aware that they are not able to set a capacity to 0, and the CDK will silently bump it to 1. The way I see it, we can go about it two ways:
> 
>     * Allow 0, warn the user
> 
>     * Document the minimum value, throw if set to 0
> 
> 
> What do you think?

I like the idea of adding to the documentation and throwing an exception. With a warning, it could easily be missed.

**Comment by nmussy** [comment]
@pkandasamy91 I also missed `ScheduledTaskBase`:

https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-ecs-patterns/lib/base/scheduled-task-base.ts#L108

Should we allow it like SQS, or throw like ALB/NLB?

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts [hunk 1]
```diff
@@ -211,7 +211,7 @@ export abstract class QueueProcessingServiceBase extends Construct {
     this.secrets = props.secrets;
 
     // Determine the desired task count (minimum) and maximum scaling capacity
-    this.desiredCount = props.desiredTaskCount || 1;
+    this.desiredCount = props.desiredTaskCount !== undefined ? props.desiredTaskCount : 1;
     this.maxCapacity = props.maxScalingCapacity || (2 * this.desiredCount);
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: One thing you should handle as well is if the desiredTaskCount is 0 then maxScalingCapacity needs to be specified and it needs to be greater than 0

Comment (nmussy) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: Done

**Pull request merged**